aten/src/ATen/TensorImpl.h

```diff
@@ -16,22 +18,26 @@ struct Tensor;
 
 namespace at {
 struct TensorImpl : public Retainable {
-  explicit TensorImpl(Type * type)
-  : is_scalar(false), type_(type) {}
+  explicit TensorImpl(Type * type, THTensor * tensor)
+  : is_scalar(false), type_(type), tensor(tensor) {}
+
+  virtual ~TensorImpl();
+
+  virtual void release_resources() override;
 
   Type & type() const {
     return *type_;
   }
   virtual const char * toString() const = 0;
-  virtual IntList sizes() const = 0;
+  virtual IntList sizes() const;
```

Comment:
can we make these non-virtual now or are SparseTensors an issue?

Reply:
SparseTensor is an issue. Once we get the size field into the base class we can kill it.